### ISSUE TITLE
Flecsi fails with lapacke errors

### ISSUE THREAD

**Issue by certik** [flecsi/flecsi#29]
On Moonlight, I am getting the following error:
```
/lustre/scratch1/turquoise/certik/repos/flecsi/src/specializations/burton/test/burton_gradients.cc:84:29: fatal error: lapacke/lapacke.h: No such file or directory
```
Which can be fixed by:
```diff
--- a/src/specializations/burton/test/burton_gradients.cc
+++ b/src/specializations/burton/test/burton_gradients.cc
@@ -81,7 +81,7 @@ protected:
 
 // this may not be a portable way to point to lapacke.h. it works on darwin.
 #include <ccomplex>
-#include <lapacke/lapacke.h>
+#include <lapacke.h>
 
 const int N = 10;
 ```
But then I get:
```
[ 18%] Linking CXX executable test/specializations/burton_gradients
/usr/bin/ld: cannot find -llapacke
collect2: error: ld returned 1 exit status
```
which I don't know how to fix, because the `third-party` didn't install any `lapacke.so`:
```
$ ls $TPL_INSTALL_PREFIX/lib
cmake            libhdf5_hl.a      libnetcdf.settings  libscotchmetis.a
libexodus.a      libhdf5.settings  libscotch.a         libszip.a
libexoIIv2for.a  libmetis.a        libscotcherr.a      libz.a
libhdf5.a        libnetcdf.a       libscotcherrexit.a  pkgconfig
```
Here are my copy & paste instructions to build on Moonlight:
```
bash
module load friendly-testing git gcc cmake/3.4.1

export TPL_INSTALL_PREFIX=$HOME/ext/flecsi-libs

git clone https://github.com/flecsi/third-party.git flecsi-thirdparty
cd flecsi-thirdparty
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=$TPL_INSTALL_PREFIX \
    -DENABLE_EXODUS=ON \
    -DENABLE_LAPACK=ON \
    -DENABLE_LAPACK=ON \
    -DENABLE_METIS=ON \
    -DENABLE_SCOTCH=ON \
    ..
make -j16


git clone --recursive -b beta-0.0-603-ge2dd9992 https://github.com/flecsi/flecsi.git
cd flecsi
mkdir build
cd build
../arch/developer-gnu
make # FAILS as indicated above
make test
```

**Comment by tuxfan** [comment]
The current setup assumes that lapacke has been installed with a package manager like yum, dnf or apt.  These tend to put the header files into a sub-directory.  John or Marc should be able to fix this.  I'll assign the issue to one of them.

**Comment by tuxfan** [comment]
Ondrej,

For the time being, just create the sub-directory and the links manually:

% cd install-prefix/include
% mkdir lapacke
% cd lapacke
% ln -s ../lapack*.h .

**Comment by wohlbier** [comment]
I sort of knew this was coming. I went back and forth on how to handle lapacke. I actually committed a change to third-party to not include lapack as a tpl, but never pushed it. Ultimately I decided to conditionally build that test with the following where LAPACK_FOUND was set by find_package(LAPACK). Though that doesn't solve the subdirectory issue either. What is really needed is find_package(LAPACKE). You can also get around it by commenting out that unit test until it is resolved.

if(LAPACK_FOUND AND ENABLE_IO)
  cinch_add_unit(burton_gradients
    SOURCES burton/test/burton_gradients.cc
    LIBRARIES flecsi ${LAPACK_LIBRARIES} ${IO_LIBRARIES}
  )
endif(LAPACK_FOUND AND ENABLE_IO)

**Comment by certik** [comment]
@tuxfan I just want to stress, that moving the `lapacke.h` is the easy issue to solve, and I already did a workaround by modifying the code itself (and I can use your workaround which doesn't require to modify the code, which is better). The hard issue is the other one I reported above, that for some reason, the `lapacke.so` is not installed.

**Comment by wohlbier** [comment]
We will get it worked out. Please hang tight.

**Comment by certik** [comment]
Thanks @wohlbier, I appreciate it. I am happy to test it once you have something.

**Comment by wohlbier** [comment]
@certik Pull down the head and try again. It should find your tpl installation and use it correctly. Let me know if it doesn't.

@tuxfan Now that I'm going through this I am remembering what a PIA lapack is. What I pushed should be considered a work around until a better solution can be implemented. In particular maybe we need a "find_package(LAPACK)" that can discern between our local TPL installation and a system installation. Presumably if a functioning system installation is found it would be preferred over our locally built one. 

System installations, as you say, usually have the /usr/include/lapacke subdirectory. In principle we could just add that directory to the include path IF FindLAPACK.cmake actually gave us a variable with it's discovered include path, but it doesn't.

Furthermore, note the exceptionally egregious hack in my update that specifically adds -lgfortan. Without this there is a linking error, and I couldn't find a cmake variable that had the fortran library in it.

Why don't you have a look and we can talk more Thursday.

**Comment by certik** [comment]
For reference, here is the commit that @wohlbier talks about: 9ba659a249e78faaedfc295a5feb824402459c87. I tested it and it works!

So this issue is fixed. Thanks @wohlbier for fixing it.

**Comment by junghans** [comment]
It seems the default is to not have `lapacke.h` in a subdirectory, see: <http://packages.ubuntu.com/wily/amd64/liblapacke-dev/filelist>

**Comment by wohlbier** [comment]
@junghans I merged your pull request. I tested it when the third-party is built without ENABLE_LAPACK, so the if branch you modified is hit. On darwin at least the system lapacke is not discovered so the test that requires it is not built. It's not "broken" per se, but it doesn't work as one would hope. I.e., one would hope when the third party libraries do not contain the lapacke build that it would find a system installation. The system installation does indeed exist on darwin but your modifications are not finding it.
